File: main/api/mcp/blockchain.py
import hashlib
import time
from main.api.utils.logging import logger

class Blockchain:
    def __init__(self):
        self.chain = []
        self.create_genesis_block()

    def create_genesis_block(self):
        genesis_block = {
            "index": 0,
            "previous_hash": "0",
            "timestamp": time.time(),
            "data": "Genesis Block",
            "nonce": 0
        }
        genesis_block["hash"] = self.calculate_hash(genesis_block)
        self.chain.append(genesis_block)

    def calculate_hash(self, block):
        block_string = f"{block['index']}{block['previous_hash']}{block['timestamp']}{block['data']}{block['nonce']}"
        return hashlib.sha256(block_string.encode()).hexdigest()

    def add_block(self, data):
        previous_block = self.chain[-1]
        new_block = {
            "index": len(self.chain),
            "previous_hash": previous_block["hash"],
            "timestamp": time.time(),
            "data": data,
            "nonce": 0
        }
        new_block["hash"] = self.calculate_hash(new_block)
        self.chain.append(new_block)
        logger.info(f"New block added: {new_block['hash']}")
        return new_block["hash"]

File: main/api/mcp/lib/blockchain.js
const crypto = require('crypto');
const logger = require('./logger');

class Blockchain {
  constructor() {
    this.difficulty = 4; // PoW difficulty
  }

  generateHash(data, nonce, previousHash) {
    return crypto.createHash('sha256')
      .update(`${data}${nonce}${previousHash}`)
      .digest('hex');
  }

  mineBlock(walletAddress, vialId, work) {
    try {
      const previousBlock = { hash: '0'.repeat(64) }; // Simplified for demo
      let nonce = 0;
      let hash;
      const data = JSON.stringify({ walletAddress, vialId, work });

      do {
        hash = this.generateHash(data, nonce, previousBlock.hash);
        nonce++;
      } while (!hash.startsWith('0'.repeat(this.difficulty)));

      const block = {
        id: `block_${Date.now()}`,
        previousHash: previousBlock.hash,
        hash,
        data: { walletAddress, vialId, work },
        nonce
      };

      logger.info(`Block mined: ${block.id} for vial ${vialId}`);
      return block;
    } catch (err) {
      logger.error(`Block mining error: ${err.message}`);
      throw err;
    }
  }

  validateBlock(block) {
    try {
      const hash = this.generateHash(
        JSON.stringify(block.data),
        block.nonce,
        block.previousHash
      );
      const isValid = hash === block.hash && hash.startsWith('0'.repeat(this.difficulty));
      logger.info(`Block validation: ${block.id} - ${isValid ? 'Valid' : 'Invalid'}`);
      return isValid;
    } catch (err) {
      logger.error(`Block validation error: ${err.message}`);
      return false;
    }
  }
}

module.exports = { Blockchain };
